Index: luci/modules/base/luasrc/model/network.lua
===================================================================
--- luci.orig/modules/base/luasrc/model/network.lua
+++ luci/modules/base/luasrc/model/network.lua
@@ -39,7 +39,7 @@ module "luci.model.network"
 
 IFACE_PATTERNS_VIRTUAL  = { }
 IFACE_PATTERNS_IGNORE   = { "^wmaster%d", "^wifi%d", "^hwsim%d", "^imq%d", "^ifb%d", "^mon%.wlan%d", "^sit%d", "^gre%d", "^lo$" }
-IFACE_PATTERNS_WIRELESS = { "^wlan%d", "^wl%d", "^ath%d", "^%w+%.network%d" }
+IFACE_PATTERNS_WIRELESS = { "^wlan%d", "^wl%d", "^ath%d", "^%w+%.network%d", "^ra%d", "^rai%d" }
 
 
 protocol = utl.class()
