Index: luci/modules/admin-full/luasrc/controller/admin/network.lua
===================================================================
--- luci.orig/modules/admin-full/luasrc/controller/admin/network.lua
+++ luci/modules/admin-full/luasrc/controller/admin/network.lua
@@ -224,6 +224,7 @@ function wifi_delete(network)
 		local dev = wnet:get_device()
 		local nets = wnet:get_networks()
 		if dev then
+			luci.sys.call("env -i /sbin/wifi down >/dev/null 2>/dev/null")
 			ntm:del_wifinet(network)
 			ntm:commit("wireless")
 			local _, net
@@ -233,7 +234,7 @@ function wifi_delete(network)
 					ntm:commit("network")
 				end
 			end
-			luci.sys.call("env -i /bin/ubus call network reload >/dev/null 2>/dev/null")
+			luci.sys.call("env -i /sbin/wifi up >/dev/null 2>/dev/null")
 		end
 	end
 
@@ -383,11 +384,12 @@ local function wifi_reconnect_shutdown(s
 	local net = netmd:get_wifinet(wnet)
 	local dev = net:get_device()
 	if dev and net then
+		luci.sys.call("env -i /sbin/wifi down >/dev/null 2>/dev/null")
 		dev:set("disabled", nil)
 		net:set("disabled", shutdown and 1 or nil)
 		netmd:commit("wireless")
 
-		luci.sys.call("env -i /bin/ubus call network reload >/dev/null 2>/dev/null")
+		luci.sys.call("env -i /sbin/wifi up >/dev/null 2>/dev/null")
 		luci.http.status(200, shutdown and "Shutdown" or "Reconnected")
 
 		return
