Index: luci/modules/admin-full/luasrc/model/cbi/admin_network/wifi.lua
===================================================================
--- luci.orig/modules/admin-full/luasrc/model/cbi/admin_network/wifi.lua
+++ luci/modules/admin-full/luasrc/model/cbi/admin_network/wifi.lua
@@ -144,6 +144,7 @@ st.ifname   = arg[1]
 local hwtype = wdev:get("type")
 local vendor = wdev:get("vendor")
 local band = wdev:get("band")
+local dev_wifimode = wdev:get("wifimode")
 local ifname = wnet:ifname()
 local netmode = wnet:mode()
 -- NanoFoo
@@ -1071,8 +1072,10 @@ cipher:depends({encryption="wpa-mixed"})
 cipher:depends({encryption="psk-mixed"})
 -- cipher:value("auto", translate("auto"))
 cipher:value("ccmp", translate("Force CCMP (AES)"))
+if dev_wifimode ~= "15" then
 cipher:value("tkip", translate("Force TKIP"))
 cipher:value("tkip+ccmp", translate("Force TKIP and CCMP (AES)"))
+end
 
 function encr.cfgvalue(self, section)
 	local v = tostring(ListValue.cfgvalue(self, section))
