Index: linux-3.10.20/fs/squashfs/xz_wrapper.c
===================================================================
--- linux-3.10.20.orig/fs/squashfs/xz_wrapper.c
+++ linux-3.10.20/fs/squashfs/xz_wrapper.c
@@ -39,8 +39,10 @@ struct squashfs_xz {
 };
 
 struct comp_opts {
-	__le32 dictionary_size;
 	__le32 flags;
+	__le16 bit_opts;
+	__le16 fp;
+	__le32 dictionary_size;
 };
 
 static void *squashfs_xz_init(struct squashfs_sb_info *msblk, void *buff,
