Index: linux-3.10.14/net/nat/hw_nat/foe_fdb.c
===================================================================
--- linux-3.10.14.orig/net/nat/hw_nat/foe_fdb.c
+++ linux-3.10.14/net/nat/hw_nat/foe_fdb.c
@@ -565,8 +565,12 @@ int _FoeDropEntry(unsigned int entry_num
 	struct FoeEntry *entry;
 	
 	entry = &PpeFoeBase[entry_num];
-	
+
+#if !defined (CONFIG_HNAT_V2)
+	entry->ipv4_hnapt.iblk2.dp = 7;
+#elif defined(CONFIG_ARCH_MT7623) || defined(CONFIG_RALINK_MT7621)
 	entry->ipv4_hnapt.iblk2.dp = 7;
+#endif
 
 #if defined (CONFIG_HNAT_V2)
 	PpeSetCacheEbl(); /*clear HWNAT cache*/
