Index: linux-3.10.14/net/nat/hw_nat/ra_nat.c
===================================================================
--- linux-3.10.14.orig/net/nat/hw_nat/ra_nat.c
+++ linux-3.10.14/net/nat/hw_nat/ra_nat.c
@@ -3575,10 +3575,12 @@ static void PpeSetDstPort(uint32_t Ebl)
 	dev=ra_dev_get_by_name("eth0");
 	printk("eth2 ifindex =%x\n", dev->ifindex);
 	DstPort[dev->ifindex] = dev;
+	DP_GMAC1 = dev->ifindex;
 #ifdef CONFIG_RAETH_GMAC2
 	dev=ra_dev_get_by_name("eth1");
 	printk("eth2 ifindex =%x\n", dev->ifindex);
 	DstPort[dev->ifindex] = dev;
+	DPORT_GMAC2 = dev->ifindex;;
 #endif
 #else
 	dev=ra_dev_get_by_name("eth2");
@@ -3962,7 +3964,7 @@ static void SetAclFwd(uint32_t Ebl)
 #else
 		WanInt = ra_dev_get_by_name("eth2.2");
 #endif
-		LanInt = ra_dev_get_by_name("br0");
+		LanInt = ra_dev_get_by_name("br-lan");
 #endif
 		for(i = 0;i< 6;i++){
 			value = RegRead(RALINK_ETH_SW_BASE + 0x2004 + (i * 0x100));
Index: linux-3.10.14/net/netfilter/nf_shortcut.c
===================================================================
--- linux-3.10.14.orig/net/netfilter/nf_shortcut.c
+++ linux-3.10.14/net/netfilter/nf_shortcut.c
@@ -125,9 +125,9 @@ int __init mtk_smb_hook_init(void)
 	struct in_ifaddr *ifa = NULL;
 
 #if LINUX_VERSION_CODE > KERNEL_VERSION(2,6,35)
-	lan_int = dev_get_by_name_rcu(&init_net, "br0");
+	lan_int = dev_get_by_name_rcu(&init_net, "br-lan");
 #else
-	lan_int = __dev_get_by_name("br0");
+	lan_int = __dev_get_by_name("br-lan");
 #endif
 	if (lan_int)
 		in_dev = __in_dev_get_rtnl(lan_int);
@@ -137,7 +137,7 @@ int __init mtk_smb_hook_init(void)
 	if (in_dev) {
 		for (ifap = &in_dev->ifa_list; (ifa = *ifap) != NULL;
 		     ifap = &ifa->ifa_next) {
-			if (!strcmp("br0", ifa->ifa_label))
+			if (!strcmp("br-lan", ifa->ifa_label))
 			{
 				lan_ifa = ifa;
 				break; /* found */
